[karthu0301] ip

karthu0301 opened the pull request:
# MaxBot - Your Personal Task Manager 🤖

> "Productivity is being able to do things that you were never able to do before." – Franz Kafka

MaxBot is your friendly and efficient chatbot that helps you manage your tasks. It is:

- **Text-based**
- **Easy to learn**
- **Blazing fast 🚀**

All you need to do is:

1. Download it from [here](https://github.com/your-repo/ip/releases).
2. Double-click it to run.
3. Add your tasks.
4. Let MaxBot manage your day! 😉

### Features:
- Manage tasks efficiently
- Manage deadlines with ease (coming soon!)
- Reminder notifications (coming soon!)

---

If you're a Java enthusiast, here's an example of its **main method**:

```java
public class Max {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

---

### Things I've done in this PR:
- [x] Implemented task management features.
- [x] Added commands for `todo`, `deadline`, and `event`.
- [x] Implemented support for date parsing.
- [ ] Add reminders for tasks.

---

**Why you'll love it:**

- It's **intuitive** and helps you stay on top of your tasks.
- Works perfectly with Java enthusiasts who want to explore their skills.
- **Best part?** It's free and open-source! 🥳

Give it a try and leave your feedback!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

karthu0301 committed:
Level-0: Rename, Greet and Exit

karthu0301 committed:
Level-1: Echo

karthu0301 committed:
Level-1: Echo

karthu0301 committed:
Level 2: Add, List

karthu0301 committed:
Level 2: Add, List

karthu0301 committed:
Level 3: Mark as Done and Not Done

karthu0301 committed:
Level 4: ToDos, Events, Deadlines

karthu0301 committed:
Level 4: ToDos, Events, Deadlines

karthu0301 committed:
A-TextUiTesting

karthu0301 committed:
Level 5: Handle Errors

karthu0301 committed:
Level 6: Delete

karthu0301 committed:
Level 6: Delete

karthu0301 committed:
A-Enums

karthu0301 committed:
tags updated

karthu0301 committed:
created branch-Level-7 and added save function

karthu0301 committed:
created Level-8, parsing date and printing

karthu0301 committed:
Add feature for Level 8: Dates and Times

karthu0301 committed:
Merge pull request #1 from karthu0301/branch-Level-8

Add feature for Level 8: Dates and Times

karthu0301 committed:
Merge branch 'branch-Level-8'

karthu0301 committed:
More OOP implemented

karthu0301 committed:
OOP code refined

karthu0301 committed:
Merge pull request #2 from karthu0301/branch-A-MoreOOP

Branch A-MoreOOP

karthu0301 committed:
organised classes into java packages

karthu0301 committed:
Merge pull request #3 from karthu0301/add-gradle-support

Add gradle support

karthu0301 committed:
gradle set up

karthu0301 committed:
JUnit set up and test

karthu0301 committed:
header comments added

karthu0301 committed:
coding standard fixes

karthu0301 committed:
coding standard fixes

karthu0301 committed:
Level-9

karthu0301 committed:
Merge pull request #4 from karthu0301/branch-A-CodingStandard

Branch a coding standard

karthu0301 committed:
Merge branch 'master' into branch-Level-9

karthu0301 committed:
Merge pull request #5 from karthu0301/branch-Level-9

Level-9

karthu0301 committed:
checkstyle added and GUI implementation started

karthu0301 committed:
Merge branch 'master' into branch-Level-10

karthu0301 committed:
Merge pull request #6 from karthu0301/branch-Level-10

checkstyle added and GUI implementation started

karthu0301 committed:
A-Varargs used

karthu0301 committed:
Merge pull request #7 from karthu0301/branch-A-Varargs

A-Varargs used

meloppeitreet commented on src/main/java/max/parser/Parser.java:
```diff
@@ -0,0 +1,50 @@
+    public static Command parse(String fullCommand) throws MaxException {
+        String[] inputParts = fullCommand.split(" ", 2);
+        String commandWord = inputParts[0];
+        String arguments = inputParts.length > 1 ? inputParts[1] : "";
```
May be good to place brackets around ternary condition?

meloppeitreet commented on src/main/java/max/task/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+
+public class Deadline extends Task {
+    private LocalDateTime by;
+    private static final DateTimeFormatter INPUT_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
I like how you followed SCREAMING_SNAKE_CASE for constants

meloppeitreet commented on src/main/java/max/task/Event.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(OUTPUT_FORMATTER) +
```
I like the line wrapping where appropriate (line length longer than 120 chars)

meloppeitreet reviewed (commented):
Overall LGTM, very nice that you have split your Command base class into individual derived classes for each type of Command

meglwhy commented on src/main/java/max/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+            throw new MaxException("Invalid task number!");
+        }
+
+        Task removedTask = tasks.deleteTask(deleteIndex);
```
This line is quite long, maybe you can break it up into multiple lines?

meglwhy commented on src/main/java/max/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+package max.command;
+
+import max.storage.Storage;
```
I like how you organized the imports here.

meglwhy commented on src/main/java/max/command/MarkCommand.java:
```diff
@@ -0,0 +1,37 @@
+        this.index = index;
+    }
+
+    /**
```
I like how you included header comments.

karthu0301 committed:
Max chatbot: add assertions to enforce critical assumptions

The codebase lacks explicit checks for important assumptions, such as
non-null task additions, valid user input, and correct parsing behavior.
This can lead to unexpected behavior or difficult-to-debug issues.

Let's add `assert` statements in key methods like `TaskList.addTask()`,
`Parser.parse()`, `Max.getResponse()`, and task constructors to ensure
essential conditions hold at runtime.

Assertions help catch logical errors early in development without
affecting production performance, as they are disabled by default.

This approach improves code robustness and maintainability. Developers
can enable assertions (`java -ea`) for debugging and disable them in
production to prevent unnecessary checks.

karthu0301 committed:
Max chatbot: refactor to improve code quality

Several areas of the codebase contain violations of best practices,
such as deep nesting, magic numbers, inconsistent naming, and redundant
comments. These issues reduce readability and maintainability.

Let's extract helper methods, remove redundant variables, use constants for
magic values, and improve Javadoc comments for better clarity.

This refactor follows the CS2103/T guidelines on code quality,
enhancing readability and maintainability while keeping the behavior
unchanged.

karthu0301 committed:
Refactor commands and task handling using Java Streams

Manual loops in FindCommand, ListCommand, and AddCommand made the
code verbose and less readable. Additionally, mutable list exposure
and index-based deletion in TaskList reduced encapsulation and
flexibility.

Let's refactor the code by:
- Using Streams (`filter()`, `collect()`) to replace explicit loops
- Leveraging `IntStream.range()` for task list formatting
- Returning `Collections.unmodifiableList()` to prevent external mutation
- Using `Arrays.stream().anyMatch()` for validation checks

These changes improve readability, remove boilerplate code, and
enhance maintainability by applying functional programming principles.

karthu0301 committed:
Refactor priority system using enums for better readability and safety

Previously, priority was managed as an integer, making it prone to
errors and difficult to understand. The system lacked clarity and
enforced priority levels using arbitrary numbers.

Let's introduce an `enum` for `Priority` and refactor:
- `Task` now stores priority as `Priority` instead of `int`.
- `AddCommand` and `PriorityCommand` updated to use enums.
- `Parser` accepts `priority low/medium/high` instead of numbers.
- `ListCommand` sorts tasks by priority using Streams.
- Ensured type safety and improved maintainability.

This refactoring improves code clarity, prevents invalid priority
values, and makes comparisons more intuitive.

karthu0301 committed:
Merge branch 'master' into branch-A-Assertions

karthu0301 committed:
Merge pull request #8 from karthu0301/branch-A-Assertions

Max chatbot: add assertions to enforce critical assumptions

karthu0301 committed:
Merge branch 'master' into branch-A-CodeQuality

karthu0301 committed:
Merge pull request #9 from karthu0301/branch-A-CodeQuality

Max chatbot: refactor to improve code quality

karthu0301 committed:
Merge branch 'master' into branch-A-Streams

karthu0301 committed:
Merge pull request #10 from karthu0301/branch-A-Streams

Refactor commands and task handling using Java Streams

karthu0301 committed:
Merge branch 'master' into branch-C-Priority

karthu0301 committed:
Merge pull request #11 from karthu0301/branch-C-Priority

Refactor priority system using enums for better readability and safety

karthu0301 committed:
Better GUI

karthu0301 committed:
GUI improved, personality added

karthu0301 committed:
More error handling implemented

karthu0301 committed:
More JUnit testing implemented

karthu0301 committed:
Merge pull request #13 from karthu0301/branch-A-BetterGUI

Better GUI

karthu0301 committed:
Resolved merge conflicts between master and branch-A-Personality

karthu0301 committed:
Merge pull request #14 from karthu0301/branch-A-Personality

GUI improved, personality added

karthu0301 committed:
Merge branch 'master' into branch-A-MoreErrorHandling

karthu0301 committed:
Merge pull request #15 from karthu0301/branch-A-MoreErrorHandling

More error handling implemented

karthu0301 committed:
Merge pull request #16 from karthu0301/branch-A-MoreTesting

More JUnit testing implemented

karthu0301 committed:
Integrated all features and fixed bugs

karthu0301 committed:
Fix Priority in database

karthu0301 committed:
Fix Checkstyle errors

karthu0301 committed:
Fix formatting issues

karthu0301 committed:
Fix priority labels and use in task classes

karthu0301 committed:
Fix warnings

karthu0301 committed:
Update README and show command logic

karthu0301 committed:
Fix test issues and update README

karthu0301 committed:
Update README and ShowCommand error handling

karthu0301 committed:
Fix final changes

karthu0301 committed:
Fix warnings

karthu0301 committed:
Fix warnings
